File: 39平衡二叉树/Solution.java
public class Solution {
    private boolean ib=true;
    public boolean IsBalanced_Solution(TreeNode root) {
        depth(root);
        return ib;
    }
    private int depth(TreeNode root){
        if (root ==null||ib==false)
            return 0;
        int left=depth(root.left);
        int right=depth (root.right);
        if (Math.abs(left-right)<=1){
            return Math.max(left,right)+1;
        }
        else {
            ib=false;
             return Math.max(left,right)+1;
        }
    }
}
/* 思路是和cpp方法一样的，但是实现不一样，这里是定义了一个全局变量ib
然后也是后续遍历，但是返回值定位了depth，这样就可以再计算depth的
同时更新ib，更新ib的计算方法和cpp是一样的 */

File: 39平衡二叉树/Solution.cpp
/* 输入一棵二叉树，判断该二叉树是否是平衡二叉树。
平衡二叉树的定义是：所谓的平衡之意，
就是树中任意一个结点下左右两个子树的高度差不超过 1。 */
class Solution {
public:
    bool IsBalanced_Solution(TreeNode* pRoot) {
        int depth = 0;
        return IsBalanced(pRoot, &depth);
    }
private:
    int IsBalanced(TreeNode* pRoot, int* depth){
        if(pRoot == NULL){
            *depth = 0;
            return true;
        }
        int left, right;
        if(IsBalanced(pRoot->left, &left) && IsBalanced(pRoot->right, &right)){
            int diff = left - right;
            if(diff <= 1 && diff >= -1){
                *depth = 1 + (left > right ? left : right);
                return true;
            }
        }
        return false;
    }
};
/* 思路是后序遍历，也就是从二叉树的叶子节点开始遍历，那么首先需要做的就是
先通过递归到达叶子节点，然后对左右树的depth进行记录，每一轮都更新一次
左右之间的差值，如果发现有超过2的则返回false */